infrastructure/comms/src/connection/net_address/net_addresses.rs

```diff
@@ -0,0 +1,253 @@
+use crate::connection::{
+    net_address::{net_address_with_stats::NetAddressWithStats, NetAddressError},
+    NetAddress,
+};
+use chrono::prelude::*;
+
+const MAX_CONNECTION_ATTEMPTS: u32 = 3;
+
+#[derive(Debug)]
+pub struct NetAddresses {
+    addresses: Vec<NetAddressWithStats>,
+}
+
+impl NetAddresses {
+    /// Constructs a new list of addresses with usage stats from a list of net addresses
+    pub fn new(addresses: Vec<NetAddressWithStats>) -> NetAddresses {
+        NetAddresses { addresses }
+    }
+
+    /// Constructs a new list of addresses with usage stats from a single net address
+    pub fn from_net_address(address: NetAddress) -> NetAddresses {
+        NetAddresses {
+            addresses: vec![NetAddressWithStats::from_net_address(address)],
+        }
+    }
+
+    /// Provides the date and time of the last successful communication with this peer
+    pub fn last_seen(&self) -> Option<DateTime<Utc>> {
+        let mut latest_valid_datetime: Option<DateTime<Utc>> = None;
+        for curr_address in &self.addresses {
+            if curr_address.last_seen.is_some() {
```

Comment:
You can unnest these ifs with
`if curr_address.last_seen.is_none() { continue; }`

Reply:
Change applied